cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaClusterCreator.java

```diff
@@ -0,0 +1,267 @@
+    }
+
+    /**
+     * Checks if the broker scale down can be done or not based on whether the nodes are empty or still have some
+     * partition-replicas assigned.
+     *
+     * @param kafkaCr   Kafka custom resource
+     * @param kafka     Kafka cluster model
+     *
+     * @return  Future with the Kafka cluster model
+     */
+    private Future<KafkaCluster> brokerScaleDownCheck(Kafka kafkaCr, KafkaCluster kafka) {
+        if (skipBrokerScaleDownCheck(kafkaCr) || kafka.removedNodes().isEmpty()) {
+            scaleDownCheckFailed = false;
+            return Future.succeededFuture(kafka);
+        } else {
+            return brokerScaleDownOperations.canScaleDownBrokers(reconciliation, vertx, kafka.removedNodes(), secretOperator, adminClientProvider)
+                    .compose(brokersContainingPartitions -> {
+                        if (!brokersContainingPartitions.isEmpty()) {
+                            LOGGER.warnCr(reconciliation, "Cannot scale down brokers {} because {} have assigned partition-replicas", kafka.removedNodes(), brokersContainingPartitions);
+                            scaleDownCheckFailed = true;
+                        } else {
+                            scaleDownCheckFailed = false;
+                        }
+
+                        return Future.succeededFuture(kafka);
+                    });
+        }
+    }
+
+    /**
+     * Reverts the broker scale down if it is not allowed because the brokers are not empty
+     *
+     * @param kafka         Instance of the Kafka cluster model that contains information needed to revert the changes
+     * @param kafkaCr       Kafka custom resource
+     * @param nodePools     List with KafkaNodePool custom resources
+     *
+     * @return  Future with KafkaAndNodePools record containing the fixed Kafka and KafkaNodePool CRs
+     */
+    private Future<KafkaAndNodePools> revertScaleDown(KafkaCluster kafka, Kafka kafkaCr, List<KafkaNodePool> nodePools)   {
+        if (nodePools == null || nodePools.isEmpty())    {
+            // There are no node pools => the Kafka CR is used
+            int newReplicasCount = kafkaCr.getSpec().getKafka().getReplicas() + kafka.removedNodes().size();
+            LOGGER.warnCr(reconciliation, "Reverting scale-down of Kafka {} by changing number of replicas to {}", kafkaCr.getMetadata().getName(), newReplicasCount);
+
+            Kafka newKafkaCr = new KafkaBuilder(kafkaCr)
+                    .editSpec()
+                        .editKafka()
+                            .withReplicas(newReplicasCount)
+                        .endKafka()
+                    .endSpec()
+                    .build();
+
+            return Future.succeededFuture(new KafkaAndNodePools(newKafkaCr, nodePools));
+        } else {
+            // Node pools are used -> we have to fix scale down in the KafkaNodePools
+            List<KafkaNodePool> newNodePools = new ArrayList<>();
+
+            for (KafkaNodePool nodePool : nodePools)    {
```

Comment:
could it make sense the nodepools with `controller` role only?

Reply:
Why would it make sense for controller role only? This is something you typically want to do for brokers. But the other changes we want to catch in the future - such as the role changes or possibly storage - would apply to all node pools.

Reply:
sorry ... I think a missed a term in my question ... I meant "skipping the node pools with controller(s) only" because they don't host topics which impact the scale down.
But I guess you already answered to it about other possible future checks.

Reply:
I guess that can be done, yeah.